File: src/main/java/org/hisp/dhis/rules/variables/ProgramRuleConstant.java
package org.hisp.dhis.rules.variables;

import org.hisp.dhis.antlr.ParserExceptionWithoutContext;
import org.hisp.dhis.rules.RuleVariableValue;
import org.hisp.dhis.rules.parser.expression.CommonExpressionVisitor;
import org.hisp.dhis.rules.parser.expression.function.ScalarFunctionToEvaluate;

import static org.hisp.dhis.parser.expression.antlr.ExpressionParser.ExprContext;

public class ProgramRuleConstant
    extends ScalarFunctionToEvaluate
{
    @Override
    public Object evaluate( ExprContext ctx, CommonExpressionVisitor visitor )
    {
        RuleVariableValue variableValue = visitor.getValueMap().get( ctx.uid0.getText() );

        if ( variableValue == null )
        {
            throw new ParserExceptionWithoutContext( "Variable " + ctx.uid0.getText() + " not present" );
        }

        return variableValue.value() == null ?
            variableValue.type().defaultValue() : getValueCastedByType( variableValue );
    }

    @Override
    public Object getDescription( ExprContext ctx, CommonExpressionVisitor visitor )
    {
        String variable = ctx.uid0.getText();

        if ( visitor.getItemStore().containsKey( variable ) )
        {
            visitor.getItemDescriptions().put( ctx.getText(), visitor.getItemStore().get( variable ).getDisplayName() );

            return visitor.getItemStore().get( ctx.uid0.getText() ).getValueType().getValue();
        }

        throw new ParserExceptionWithoutContext(
            "Variable " + ctx.uid0.getText() + " does not exist" );
    }
}

File: src/jvmMain/kotlin/org/dhis2/ruleengine/variables/ProgramRuleConstant.kt
package org.dhis2.ruleengine.variables

import org.dhis2.ruleengine.parser.expression.CommonExpressionVisitor
import org.dhis2.ruleengine.parser.expression.function.ScalarFunctionToEvaluate
import org.hisp.dhis.antlr.ParserExceptionWithoutContext
import org.hisp.dhis.parser.expression.antlr.ExpressionParser.ExprContext

class ProgramRuleConstant : ScalarFunctionToEvaluate() {
    override fun evaluate(ctx: ExprContext, visitor: CommonExpressionVisitor): Any {
        val variableValue = visitor.getValueMap()[ctx.uid0.text]
            ?: throw ParserExceptionWithoutContext("Variable " + ctx.uid0.text + " not present")
        return if (variableValue.value == null) variableValue.ruleValueType.defaultValue else variableValue.value!!
    }

    override fun getDescription(ctx: ExprContext, visitor: CommonExpressionVisitor): Any {
        val variable = ctx.uid0.text
        if (visitor.getItemStore().containsKey(variable)) {
            visitor.itemDescriptions[ctx.text] = visitor.getItemStore()[variable]!!.displayName
            return visitor.getItemStore()[ctx.uid0.text]!!.valueType.value
        }
        throw ParserExceptionWithoutContext(
            "Variable " + ctx.uid0.text + " does not exist"
        )
    }
}